fdbserver/ClusterController.actor.cpp

```diff
@@ -3344,6 +3355,27 @@ struct BlobManagerSingleton : Singleton<BlobManagerInterface> {
 	void recruit(ClusterControllerData* cc) const { cc->recruitBlobManager.set(true); }
 };
 
+struct EncryptKeyProxySingleton : Singleton<EncryptKeyProxyInterface> {
```

Comment:
Given that this process is required to be available for all stateful roles to be fully available and able to respond to reads+make progress if data is encrypted, it seems like having it as a singleton is potentially a big availability bottleneck. Do we have plans to recruit more than one of these down the line, and distribute the requests across them?

Reply:
Agree that _singleton_ process may need a reliability boost by recruiting _multiple_ processes; I was hoping that monitoring _single interface_ with tighter failure time detection may work for now; but given _multi process_ is the goal, better to make changes now. 

Ideas about the _multi process_ transformation:
1. Configurable number of process count.
2. Part of recovery: I don't think EKP need to be part of recovery; especially given _encryption key cache_ can live across recoveries and is a more efficient solution. 

Proposal is to let CC recruit multiple processes as it is doing for _singleton_ and implement an actor to monitor recruited interfaces; if they go down, recruit new worker to satisfy the desired process count. 

Thanks.

Reply:
Thinking more "multi process" also has downsides, comparing two scheme below:
1. _Singleton_ has reliability concerns, however, a given `encryption key` will be fetched _only_ once from GS KeyManager. 
Is setting a _failure detection_ timer to a smaller value such as _50-100_ ms good enough? Will such an interval good enough to avoid any _false_ process recovery due to flaky network?

2. _Multi process_ approach adds more reliability; the code can associate EKP processes with CC lifecycle and monitor multiple interfaces for failure to ensure enough availability.
The downside with this approach is: _encryption key_ could be fetched multiple times due to request load-balancing; we can mitigate the issue, however, at the cost of added code complexity.
a. Key Sharding to associate a key-range to EKP process.
b. Gossip "hot keys" - need to define key-temperature though.
c. ..More scheme possible for sure..

Thoughts?
cc: @etschannen

Reply:
The main concern with it being a singleton would be if it is a performance bottleneck. Ratekeeper and the sequencer are already singleton stateless processes which stop a cluster if they are not running.

Reply:
Agree about the "performance" bottleneck given EKP would be accessed by all other processes including SS, BlobWorkers etc. 

I'll update the code to make it a multi-process role (configurable count); as proposed earlier, will tie its life-cycle with CC (not with cluster recovery) and add an actor to monitor recruited interfaces. 

Performance optimization: inorder to mitigate multiple same encryption-key fetches; any new key fetched from GS can be broadcasted to all EKP processes (separate gossip endpoint), depending on usecase, this gossiping of keys should avoid multiple GS round-trips (this optimization can be done later as well, I think)

Reply:
Summarizing offline discussion with @sfc-gh-etschannen. 

EKP process would fetch _encryption keys_ from GS, however, it would be useful to avoid per-key lookup from other FDB processes such as: TLog/CP, SSs, BlobWorkers etc.; given the _encryption key_ is immutable, other processes can cache them locally. 
EKP process will periodically refresh the _latest active key_ from GS and on an update can either broadcast keys to other interested FDB processes, or we can have a cache TTL, which forces the key refresh by fetching keys again from EKP (on demand). 

Based on above discussion, the PR proposal of making EKP a _singleton_ seems suitable. Thanks.